File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.TreeMap;

class Solution {
    
    public int[] solution(String[] operations) {
        TreeMap<Integer, Integer> map = new TreeMap<>();

        for (String operation : operations) {
            String[] tokens = operation.split(" ");
            String command = tokens[0];
            int value = Integer.parseInt(tokens[1]);

            if (command.equals("I")) {
                map.put(value, map.getOrDefault(value, 0) + 1);
                continue;
            }
            if (!map.isEmpty()) {
                int key = (value == 1)
                        ? map.lastKey()
                        : map.firstKey();

                int count = map.get(key);

                if (count == 1) {
                    map.remove(key);
                } else {
                    map.put(key, count - 1);
                }
            }
        }

        if (map.isEmpty()) {
            return new int[]{0, 0};
        }

        return new int[]{map.lastKey(), map.firstKey()};
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heappush, heappop, heapify, nsmallest


def solution(operations):
    max_heap = []
    min_heap = []
    count = 0
    
    for operation in operations:
        op, num_str = operation.split()
        num = int(num_str)
        
        if op == 'I':
            heappush(min_heap, num)
            heappush(max_heap, -num)
            count += 1
            
        if count == 0:
            continue
        
        if op == 'D':
            if num == 1:
                heappop(max_heap)
                min_heap = nsmallest(len(max_heap), min_heap)
                heapify(min_heap)
                count -= 1
            elif num == -1:
                heappop(min_heap)
                max_heap = nsmallest(len(min_heap), max_heap)
                heapify(max_heap)
                count -= 1
                
            if count == 0:
                max_heap.clear()
                min_heap.clear()
    
    if count == 0:
        return 0, 0
    
    return -max_heap[0], min_heap[0]
